### ISSUE TITLE
Narrowing of a Union containing an Enum doesn't work

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#1309]
Here, `x` is reported to have the type `MyEnum | str` in all three branches:
```py
@final
class MyEnum(Enum):
    ZERO = 0
    ONE = 1


def f(x: Union[MyEnum, str]):
    if x is MyEnum.ZERO:
        x
    elif x is MyEnum.ONE:
        x
    else:
        x
```
using `==` instead of `is` yields the same result.

Here, however, all branches have the right type inferred:
```py
def g(x: Union[MyEnum, str]):
    if isinstance(x, MyEnum):
        if x is MyEnum.ZERO:
            x  # Literal[MyEnum.ZERO]
        elif x is MyEnum.ONE:
            x  # Literal[MyEnum.ONE]
    else:
        x  # str
```

**Comment by erictraut** [comment]
The previous logic intentionally backed off all type narrowing in this more complicated case involving a union with enum and non-enum types. This strikes me as a really odd usage, so I was tempted to mark the issue as "as designed" and close it. But it's relatively easy to make the narrowing logic a little smarter in this case, so I've gone ahead and added support for this unusual case.

**Comment by decorator-factory** [comment]
I was thinking of using this for defining sentinels (aka `object()`) that are supposed to escape the function: https://github.com/microsoft/pyright/discussions/1311
Although... I could've just used a `str` and let it infer a literal type.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.97, which I just published. It will also be included in the next release of pylance.
